File: boligportal.dk/worker.go
package main

import (
  "encoding/json"
  "fmt"
  "log"
  "os"

  "github.com/streadway/amqp"
)

const OffersQueue = "offers"
const BoligPortalSearchsQueue = "boligportal_searchs"

func crawlOffers(channel *amqp.Channel) {
  offers, err := Crawl("https://www.boligportal.dk/RAP/ads?placeIds[]=15&housingTypes[]=3&listViewResults=true&limitRecords=60&sort=paid&tid=5c0283e94d35b")
  if err != nil {
    log.Fatalf("Could not crawl offers: %s", err)
  }

  for _, offer := range offers {
    jsonOffer, err := json.Marshal(offer)
    if err != nil {
      log.Fatalf("Could not marshal offer: %s", err)
    }

    fmt.Printf("%s -- %s\n", offer.Title, offer.Identifier)

    err = channel.Publish(
      "",     // exchange
      OffersQueue, // routing key
      false,  // mandatory
      false,  // immediate
      amqp.Publishing {
        ContentType: "application/json",
        Body: jsonOffer,
      },
    )
    if err != nil {
      log.Fatalf("Could not publish offer: %s", err)
    }
  }
}

func main() {
  rabbitMqUrl, ok := os.LookupEnv("RABBITMQ_URL")
  if !ok {
    log.Fatal("The DSN to use to connect to RabbitMq must be specified by the RABBITMQ_URL environment variable.")
  }

  rabbitConn, err := amqp.Dial(rabbitMqUrl)
  if err != nil {
    log.Fatalf("Could not connect to RabbitMq: %s", err)
  }
  defer rabbitConn.Close()

  channel, err := rabbitConn.Channel()
  if err != nil {
    log.Fatalf("Could not connect to RabbitMq channel: %s", err)
  }
  defer channel.Close()

  _, err = channel.QueueDeclare(
    OffersQueue, // name
    true,   // durable
    false,   // delete when unused
    false,   // exclusive
    false,   // no-wait
    nil,     // arguments
  )
  if err != nil {
    log.Fatalf("Could not declare RabbitMq offers queue: %s", err)
  }

  _, err = channel.QueueDeclare(
    BoligPortalSearchsQueue, // name
    true,   // durable
    false,   // delete when unused
    false,   // exclusive
    false,   // no-wait
    nil,     // arguments
  )
  if err != nil {
    log.Fatalf("Could not declare RabbitMq search requests queue: %s", err)
  }

  msgs, err := channel.Consume(
    BoligPortalSearchsQueue, // queue
    "",     // consumer
    true,   // auto-ack
    false,  // exclusive
    false,  // no-local
    false,  // no-wait
    nil,    // args
  )
  if err != nil {
    log.Fatalf("Could not register search offers consumer: %s", err)
  }

  forever := make(chan bool)

  go func() {
    for _ = range msgs {
      log.Printf("Received search request!")

      crawlOffers(channel)
    }
  }()

  log.Printf("[*] Waiting for search requests. To exit press CTRL+C")
  <-forever
}

File: leboncoin/worker.py
#!/usr/bin/env python3

import crawler
import json
import logging
import os
import pika
import sys

OFFERS_QUEUE = 'offers'
LEBONCOIN_SEARCHS_QUEUE = 'leboncoin_searchs'

def search_leboncoin(rabbit_channel):
    leboncoin = crawler.Crawler()

    for offer in leboncoin.offers():
        logging.info('Found offer "%s" -- %s', offer['title'], offer['identifier'])
        rabbit_channel.basic_publish(exchange='', routing_key=OFFERS_QUEUE, body=json.dumps(offer), properties=pika.BasicProperties(
            delivery_mode = 2, # make message persistent
        ))

def search_request_received(channel, method, properties, body):
    logging.info('Received search request!')

    search_leboncoin(channel)

if __name__ == '__main__':
    logging.basicConfig(format='%(asctime)s - %(message)s', level=logging.INFO)
    rabbitMqUrl = os.environ.get('RABBITMQ_URL')

    if rabbitMqUrl is None:
        print("The DSN to use to connect to RabbitMq must be specified by the RABBITMQ_URL environment variable.", file=sys.stderr)
        sys.exit(1)

    rabbitmq = pika.BlockingConnection(pika.URLParameters(rabbitMqUrl))
    channel = rabbitmq.channel()
    channel.queue_declare(queue=OFFERS_QUEUE, durable=True)
    channel.queue_declare(queue=LEBONCOIN_SEARCHS_QUEUE, durable=True)

    logging.info('Waiting for search requests…')
    channel.basic_consume(search_request_received, queue=LEBONCOIN_SEARCHS_QUEUE)

    try:
        channel.start_consuming()
    except KeyboardInterrupt:
        rabbitmq.close()
